### PR TITLE
Adding tree walk 

### PR DESCRIPTION
**Pull request by iptq**
Just wanted to get some code down and start some discussion. I haven't worked much with bindings before so I'd appreciate some tips.

#267

### CODE REVIEW

**Comment by iptq** [comment]
What should the callback function return? I actually only put bool cuz the filter function I referenced did but doesn't really make sense in this context.

**Comment by alexcrichton** [comment]
According to libgit2:

> If the callback returns a positive value, the passed entry will be skipped on the traversal (in pre mode). A negative value stops the walk.

Could errors translate to -1 and the boolean be used to skip/not skip?

**Comment by neithernut** [comment]
Libgit2 exposes a specific error number for such cases (stopping an iteration):
https://github.com/libgit2/libgit2/blob/9913e2df077fd9ed4c0fd5f81075ca5a153cd34b/include/git2/errors.h#L34

**Comment by iptq** [comment]
How about using a custom enum type that's kind of like Option but also includes an option to stop? Seems like there's three kinds of actions that could take place here:

```rs
enum TreeWalkResult {
    Ok,
    Skip,
    Stop,
}
```

Not sure about the exact naming but then we could implement either `From<TreeWalkResult> for i32` or `Into<i32> for TreeWalkResult` and convert it to an integer for git_tree_walk.

Thoughts?

**Comment by iptq** [comment]
@alexcrichton what are your thoughts on this implementation?

File: src/tree.rs [hunk 1]
```diff
@@ -119,6 +172,27 @@ impl<'repo> Tree<'repo> {
     }
 }
 
+type TreeWalkCb<'a, T> = FnMut(&str, &TreeEntry) -> T + 'a;
+
+extern fn treewalk_cb(root: *const c_char, entry: *const raw::git_tree_entry, payload: *mut c_void) -> c_int {
+    match panic::wrap(|| unsafe {
+        if panic::panicked() {
```
[the remark this answers is not in the document]
Comment (iptq) [reply src/tree.rs]: ok :+1:

**Comment by alexcrichton** [comment]
Thanks!

**Comment by asayers** [comment]
I'm seeing some strange behaviour with this feature.  I have a callback which returns `TreeWalkResult::Ok` unconditionally, and I expect every node in the tree to be visited. Instead I see a subset of the top-level nodes only, and this subset is not deterministic across runs! Looks like something nasty is happening.  Returning `0` from the callback instead gives the expected behaviour.

**Comment by iptq** [comment]
That's not good! I'll take a look after classes today.

**Comment by asayers** [comment]
Also, reading https://github.com/libgit2/libgit2/pull/4703 makes it seem like we should be passing `GIT_EUSER` (-7) rather than -1 to abort the traversal. (Unrelated to previous comment.)

**Pull request merged**